### PR TITLE
chore: Add Destintaion Name to Middleware

### PR DESCRIPTION
**Pull request by FrankEssenberger**
realtes to https://github.com/SAP/cloud-sdk-js/issues/3315

Adds the destinationName to the middleware interface so that users can refetch destinations if needed.

### CODE REVIEW

File: packages/http-client/src/http-client.spec.ts [hunk 1]
```diff
@@ -305,6 +319,77 @@ describe('generic http client', () => {
...
+        jwt: subscriberUserJwt,
+        requestConfig: expect.any(Object),
+        tenantId: 'subscriber',
+        uri: 'https://my.system.com'
+      });
+
+      delete process.env['VCAP_SERVICES'];
+      nock.cleanAll();
+      jest.clearAllMocks();
+    }, 999999);
```
Comment (florian-richter) [inline packages/http-client/src/http-client.spec.ts:391]: [q] Is this timeout length necessary?

File: packages/http-client/src/http-client.spec.ts [hunk 2]
```diff
@@ -305,6 +319,77 @@ describe('generic http client', () => {
...
+          middleware: [showContextMiddleware],
+          method: 'get'
+        }
+      );
+      expect(response.data).toEqual({
+        destinationName: 'FINAL-DESTINATION',
+        jwt: subscriberUserJwt,
+        requestConfig: expect.any(Object),
+        tenantId: 'subscriber',
+        uri: 'https://my.system.com'
```
Comment (florian-richter) [inline packages/http-client/src/http-client.spec.ts:385]:
```suggestion
        uri: 'https://my.system.example.com'
```
[req] Always use example.com or an SAP URL

File: test-resources/test/test-util/mocked-access-tokens.ts [hunk 3]
```diff
@@ -2,7 +2,7 @@ import {
   onlyIssuerXsuaaUrl,
   providerXsuaaUrl,
   subscriberXsuaaUrl,
-  TestTenants
+  TestTenants, xsuaaBindingMock
```
Comment (florian-richter) [inline test-resources/test/test-util/mocked-access-tokens.ts:5]:
```suggestion
  TestTenants, 
  xsuaaBindingMock
```
[req] formatting

File: test-resources/test/test-util/mocked-access-tokens.ts [hunk 4]
```diff
@@ -89,7 +89,10 @@ export const subscriberUserPayload = {
   iat,
   iss: subscriberXsuaaUrl,
   zid: TestTenants.SUBSCRIBER,
-  user_id: 'user-sub'
+  user_id: 'user-sub',
+  //Becomes clientId in xssec
+  azp: xsuaaBindingMock.credentials.clientid,
+  aud: [xsuaaBindingMock.credentials.clientid]
```
Comment (florian-richter) [inline test-resources/test/test-util/mocked-access-tokens.ts:95]: [q] Could you add some context why these are needed? (Maybe not in this PR but in the docs intead...)

**Review by florian-richter — CHANGES_REQUESTED** [review]

File: test-resources/test/test-util/mocked-access-tokens.ts (hunk 4 above)
Comment (FrankEssenberger) [reply test-resources/test/test-util/mocked-access-tokens.ts:95]: This is needed when XSSEC validates the JWT, which is mocked in general but for the test along packages I could not mock it with jest. Do you have a place in mind?

Comment (florian-richter) [reply test-resources/test/test-util/mocked-access-tokens.ts:95]: Maybe add a comment above if it just for testing.

Comment (FrankEssenberger) [reply test-resources/test/test-util/mocked-access-tokens.ts]: extended the comment

**Comment by fwilhe** [comment]
@florian-richter are your reasons to request changes still valid?

File: packages/http-client/src/http-client.ts [hunk 5]
```diff
@@ -133,6 +135,7 @@ export function execute(executeFn: ExecuteHttpRequestFn<HttpResponse>) {
         jwt: destination.jwt,
         requestConfig: request,
         uri: resolvedDestination.url,
+        destinationName: resolvedDestination.name ?? undefined,
```
Comment (fwilhe) [inline packages/http-client/src/http-client.ts:138]: Not fully clear to me how this is different from omitting the `?? undefined` part. Is this about null vs undefined?

**Review by fwilhe — APPROVED** [review]
LGTM

File: packages/http-client/src/http-client.ts (hunk 5 above)
Comment (florian-richter) [reply packages/http-client/src/http-client.ts:138]: Seems like that has to be the reason... but I am surprised that `resolvedDestination.name` would ever be `null`

File: test-resources/test/test-util/mocked-access-tokens.ts [hunk 6]
```diff
@@ -89,7 +90,10 @@ export const subscriberUserPayload = {
   iat,
   iss: subscriberXsuaaUrl,
   zid: TestTenants.SUBSCRIBER,
-  user_id: 'user-sub'
+  user_id: 'user-sub',
+  // The client and audience are necesssary if XSSEC validates the token
+  azp: xsuaaBindingMock.credentials.clientid,   //Becomes clientId in XSSEC
+  aud: [xsuaaBindingMock.credentials.clientid]  //Becomes audience in XSSEC
```
Comment (florian-richter) [inline test-resources/test/test-util/mocked-access-tokens.ts:96]:
```suggestion
  azp: xsuaaBindingMock.credentials.clientid,   // Becomes clientId in XSSEC
  aud: [xsuaaBindingMock.credentials.clientid]  // Becomes audience in XSSEC
```
[nth]

**Review by florian-richter — APPROVED** [review]

File: packages/http-client/src/http-client.ts (hunk 5 above)
Comment (FrankEssenberger) [reply packages/http-client/src/http-client.ts:138]: yes, I wanted to remove the `null` in the types of the destination but I decided not to do that in this PR. Perhaps the destination service returns objects with `null` value in the JSON?

**Pull request merged**